common/src/main/java/io/netty/util/concurrent/DefaultFutureListeners.java

```diff
@@ -15,53 +15,75 @@
  */
 package io.netty.util.concurrent;
 
+import io.netty.util.internal.logging.InternalLogger;
+
 import java.util.Arrays;
+import java.util.EventListener;
 
 final class DefaultFutureListeners {
-
-    private GenericFutureListener<? extends Future<?>>[] listeners;
+    private Object[] listeners;
     private int size;
 
-    @SuppressWarnings("unchecked")
-    DefaultFutureListeners(
-            GenericFutureListener<? extends Future<?>> first, GenericFutureListener<? extends Future<?>> second) {
-        listeners = new GenericFutureListener[2];
-        listeners[0] = first;
-        listeners[1] = second;
-        size = 2;
+    DefaultFutureListeners() {
+        listeners = new Object[4];
     }
 
-    public void add(GenericFutureListener<? extends Future<?>> l) {
-        GenericFutureListener<? extends Future<?>>[] listeners = this.listeners;
-        final int size = this.size;
-        if (size == listeners.length) {
+    public void add(EventListener listener, Object context) {
+        Object[] listeners = this.listeners;
+        int size = this.size;
+        if (size == listeners.length || listener instanceof FutureContextListener && size == listeners.length - 1) {
             this.listeners = listeners = Arrays.copyOf(listeners, size << 1);
         }
-        listeners[size] = l;
-        this.size = size + 1;
+        listeners[size++] = listener;
+        if (listener instanceof FutureContextListener) {
+            listeners[size++] = context;
+        }
+        this.size = size;
     }
 
-    public void remove(GenericFutureListener<? extends Future<?>> l) {
-        final GenericFutureListener<? extends Future<?>>[] listeners = this.listeners;
+    public void remove(EventListener listener) {
+        final Object[] listeners = this.listeners;
         int size = this.size;
         for (int i = 0; i < size; i ++) {
```

Comment:
This isn't right, as we need to double-step if the current entry is a `FutureContextListener` as we don't want to compare against a Context.